Expose the task object in SnapshotTaskInputsOperationDetails

alkemist opened the pull request:
Needed for build scans.

oehme commented on subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsOperationDetails.java:
```diff
@@ -28,20 +30,19 @@
 
-    public String getTaskPath() {
-        return taskPath;
+    public Task getTask() {
```
We shouldn't leak mutable state into operation details. Can you instead expose the details you need from the task?

facewindu replied:
We need something that is guaranteed to be unique (and the task path isn't), in order to correctly link TaskStarted/TaskFinished events (created when our TaskExecutionListener is called in the build scan plugin) and the new event created when listening on this SnapshotTaskInputsOperationDetails.
But I'll let @alkemist expose further.

oehme replied:
`TaskInternal#getIdentityPath` is unique

alkemist replied:
@oehme Task identity paths are not guaranteed to be unique. Gradle will happily execute two tasks of the same project with the same name. This is why we have to key off the task object. It's the only thing that's guaranteed to be unique. 

Try this build script and run `all`:

```
task all

task t {
  doFirst {
    println "first"
  }
}

all.dependsOn t

task t(overwrite: true) {
  doFirst {
    println "second"
  }
}

task intermediary {
  dependsOn t
}

all.dependsOn intermediary
```

This is not as uncommon as you might think, and caused us problems at the start of build scans.

oehme replied:
This is terrible, we should fix that, since it's obviously not what the user intended. I'm sure we have plenty of code that assumes `Task#identityPath` to be unique.

That being said, ff there's no other way for now than to expose the `Task`, I'm okay with that.

oehme replied:
I just talked with @adammurdoch about this: You could use the identity hashCode instead (which will be unique since we hold on to both tasks for the full duration of the build). It would be really great if we could avoid leaking the task object. We definitely can't commit to the `Task` being the contract long term.

oehme commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/BuildOperationNotificationsFixture.groovy:
```diff
@@ -1,131 +0,0 @@
-    }
-
-    @Override
-    String initScriptContent() {
```
What was wrong with the init script approach? The new way seems pretty hacky, requiring a bunch of new types in production code which is actually only there for testing.

alkemist replied:
It's clumsy and error prone, and makes it difficult to augment the serialization where necessary.

oehme replied:
> The implementation can be cleaned up considerably. 

Please do so. This is some of our most important infrastructure, so we need it to be very clear.

> I had to get it working very quickly.

We have another week until RC1. Even if we didn't, I'd rather postpone RC1 than rush changes to central infrastructure.

facewindu commented:
@alkemist there are [some failing tests](https://e.grdev.net/s/2c2nq7zrhbrhs/tests/failed).

alkemist commented:
@oehme I reworked this to not require changing the build operations test fixture, and moved those changes to https://github.com/gradle/gradle/pull/2033.

facewindu commented:
@oehme please confirm you're OK with the PR as it is now

oehme commented:
I think we can avoid leaking the task object by using the identity hash. Ideally, we'd fix the task replacement problem instead of working around it.

alkemist commented:
@oehme `identityHashCode` is not guaranteed to be unique.

oehme commented:
@alkemist the identity hash uses the memory address of the object. Since the tasks can't be GC'd while the build executes, no two tasks in the same execution will have the same identity hash.
